pkg/distributor/otel.go

```diff
@@ -414,33 +430,70 @@ func otelMetricsToMetadata(addSuffixes bool, md pmetric.Metrics) []*mimirpb.Metr
 	return metadata
 }
 
-func otelMetricsToTimeseries(ctx context.Context, tenantID string, addSuffixes, enableCTZeroIngestion, enableStartTimeQuietZero bool, promoteResourceAttributes []string, keepIdentifyingResourceAttributes bool, discardedDueToOtelParseError *prometheus.CounterVec, logger log.Logger, md pmetric.Metrics) ([]mimirpb.PreallocTimeseries, error) {
-	converter := otlp.NewMimirConverter()
-	_, errs := converter.FromMetrics(ctx, md, otlp.Settings{
+func otelMetricsToTimeseries(
+	ctx context.Context,
+	converter *otlpMimirConverter,
+	addSuffixes, enableCTZeroIngestion, enableStartTimeQuietZero bool,
+	promoteResourceAttributes []string,
+	keepIdentifyingResourceAttributes bool,
+	md pmetric.Metrics,
+	logger log.Logger,
+) ([]mimirpb.PreallocTimeseries, int, error) {
+	settings := otlp.Settings{
 		AddMetricSuffixes:                   addSuffixes,
 		EnableCreatedTimestampZeroIngestion: enableCTZeroIngestion,
 		EnableStartTimeQuietZero:            enableStartTimeQuietZero,
 		PromoteResourceAttributes:           promoteResourceAttributes,
 		KeepIdentifyingResourceAttributes:   keepIdentifyingResourceAttributes,
-	}, utillog.SlogFromGoKit(logger))
-	mimirTS := converter.TimeSeries()
-	if errs != nil {
-		dropped := len(multierr.Errors(errs))
-		discardedDueToOtelParseError.WithLabelValues(tenantID, "").Add(float64(dropped)) // Group is empty here as metrics couldn't be parsed
-
-		parseErrs := errs.Error()
-		if len(parseErrs) > maxErrMsgLen {
-			parseErrs = parseErrs[:maxErrMsgLen]
-		}
+	}
+	mimirTS := converter.ToTimeseries(ctx, md, settings, logger)
 
-		if len(mimirTS) == 0 {
-			return nil, errors.New(parseErrs)
-		}
+	dropped := converter.DroppedTotal()
+	if len(mimirTS) == 0 && dropped > 0 {
+		return nil, dropped, converter.Err()
+	}
+	return mimirTS, dropped, nil
+}
 
-		level.Warn(logger).Log("msg", "OTLP parse error", "err", parseErrs)
+type otlpMimirConverter struct {
+	converter *otlp.MimirConverter
+	// err holds OTLP parse errors
+	err error
+}
+
+func newOTLPMimirConverter() *otlpMimirConverter {
+	return &otlpMimirConverter{
+		converter: otlp.NewMimirConverter(),
 	}
+}
 
-	return mimirTS, nil
+func (c *otlpMimirConverter) ToTimeseries(ctx context.Context, md pmetric.Metrics, settings otlp.Settings, logger log.Logger) []mimirpb.PreallocTimeseries {
+	if c.err != nil {
+		return nil
+	}
+
+	_, c.err = c.converter.FromMetrics(ctx, md, settings, utillog.SlogFromGoKit(logger))
+
+	mimirTS := c.converter.TimeSeries()
+	return mimirTS
```

Comment:
[Nit] Easier to read:
```suggestion
	return c.converter.TimeSeries()
```